Add standalone scip-ruby Docker image for CI/CD

jupblb opened the pull request:
- Split Docker configuration into three files:
  - Dockerfile.base: Common dependencies and scip-ruby binary
  - Dockerfile: Standard scip-ruby image for CI/CD pipelines
  - Dockerfile.autoindex: Autoindex variant with additional script
- Update GitHub Actions workflow to build base image locally (not published)
- Both final images (scip-ruby and autoindex) are published to Docker Hub
- New image can be used in CI: docker run -v $(pwd):/sources sourcegraph/scip-ruby

Amp-Thread-ID: https://ampcode.com/threads/T-93250ea7-f00b-492f-ba56-04da7ddfe7d9

### Motivation
We currently publish autoindex images on Docker Hub. We could use an image with `scip-ruby` binary, for example in a continuous integration job.


### Test plan
Checked manually if everything builds and works as expected

Edit: Trying to rerun the checks

jupblb committed:
Add standalone scip-ruby Docker image for CI/CD

- Split Docker configuration into three files:
  - Dockerfile.base: Common dependencies and scip-ruby binary
  - Dockerfile: Standard scip-ruby image for CI/CD pipelines
  - Dockerfile.autoindex: Autoindex variant with additional script
- Update GitHub Actions workflow to build base image locally (not published)
- Both final images (scip-ruby and autoindex) are published to Docker Hub
- New image can be used in CI: docker run -v $(pwd):/sources sourcegraph/scip-ruby

Amp-Thread-ID: https://ampcode.com/threads/T-93250ea7-f00b-492f-ba56-04da7ddfe7d9
Co-authored-by: Amp <amp@ampcode.com>

varungandhi-src commented on Dockerfile.base:
```diff
@@ -0,0 +1,35 @@
+# Use a release binary instead of building from source
+# because release builds are very time-consuming.
+#
+# The release version is verified by tools/scripts/publish-scip-ruby.sh
+# The binary at this version is guaranteed to exist because this workflow runs either:
+# 1. Manually with a specific tag that already has a release
+# 2. After the Release workflow completes (which creates the release and uploads the binary)
+RUN wget https://github.com/sourcegraph/scip-ruby/releases/download/scip-ruby-v${SCIP_RUBY_VERSION}/scip-ruby-x86_64-linux \
```
This is going to break, because the version number is no longer hard-coded in the URL, so the publishing script cannot cross-check this.

jupblb replied:
I'm not sure I follow. The publishing step building this image sets `SCIP_RUBY_VERSION=${{ env.PATCH }}`.

In case this value is not passed we can return an error:
```
RUN test -n "${SCIP_RUBY_VERSION}" || \
    (echo "ERROR: SCIP_RUBY_VERSION build arg is required" && exit 1)
```

varungandhi-src reviewed (commented):
I'm curious as to what this is buying us. It seems overall, this introduces more points of failure, a bit more indirection, and the size between the images is not that different. The main difference is one extra shell script. So I'm thinking it'll be simpler to use the autoindexing image directly downstream.

Do you see it differently? I don't mind having this change, but it's unclear as to what the benefit is. Is it that we don't have to hard-code a different image tag in the downstream workflow? Is there some other benefit as well.

jupblb commented:
The original image can also be used to just run `scip-ruby`. This PR is not necessary to make use of the binary. The benefits here are more related to readability and ease of use:

1)  Push an image with `latest` tag, in line with other `scip-*` images we provide (this is a popular convention).
2)  `autoindex-*` image suggests that this image should be used to index an existing gem, not locally checked out source-code (although this is supported by `autoindex.sh`).
3)  Currently, the default binary is plain `sh`, not `scip-ruby`.
4)  Moreover, in the `Dockerfile` we have a comment saying:

    ```
    # This Docker image is meant for auto-indexing support in Sourcegraph
    # and is not recommended for third-party use.
    ```

jupblb committed:
Add validation for SCIP_RUBY_VERSION build arg in Dockerfile.base

Amp-Thread-ID: https://ampcode.com/threads/T-439e5735-e594-4e46-a41a-ca91575956a7
Co-authored-by: Amp <amp@ampcode.com>

varungandhi-src reviewed (approved):
OK with adding a separate image.

Nice to have but not necessary: Adding a workflow so that we're testing the build of the image when making PRs (I think scip-python has something similar).

jupblb commented:
> Nice to have but not necessary: Adding a workflow so that we're testing the build of the image when making PRs (I think scip-python has something similar).

We do have a similar workflow to that in [scip-python](https://github.com/sourcegraph/scip-python/blob/scip/.github/workflows/scip-docker.yml) but its usefulness stems from the fact that it also builds the project. In this workflow we use an already published binary. There is no reason to run it on each push as it would have the same effect.
